### PR TITLE
Upgrade python chess library to latest version (1.4.0)

### LINKED ISSUE #240: python-chess library obsolete
The wrapper is using python chess v. 0.24, while the latest is 1.3!
In my case the problem is that my engine use also python chess and the two versions are incompatible (I've just recently updated to last version).
I think that supporting the newer versions should be a welcomed approach to program maintenance.

### PR DESCRIPTION
**Pull request by MarkZH**
This commit updates lichess-bot to use the latest version of the python
chess library. The main result is a large decrease in the amount of code
in engine_wrapper.py due to so many functions being incorporated into
the chess library.

One annoying difference is that the new library rejects the 'usermove'
feature from Xboard engines, which is rare among chess engine
interfaces.

I hope this can be tested by a variety of engines to make sure it is on
par with the current lichess-bot. This would close #240.

### CODE REVIEW

**Comment by AttackingOrDefending** [comment]
My bot crashed. It had 2 problems. https://github.com/MarkZH/lichess-bot/blob/python-chess-upgrade/engine_wrapper.py#L77 because as it writes https://python-chess.readthedocs.io/en/v1.3.3/engine.html#chess.engine.Protocol.configure you should not set options that are managed automatically (UCI_Chess960, UCI_Variant, MultiPV, Ponder) which you can see https://python-chess.readthedocs.io/en/v1.3.3/engine.html#chess.engine.Option.is_managed. I added a simple fix that probably shouldn't be used because I don't know if I can change these 4 parameters.
 `        try:
            options.pop('Ponder')
        except:
            pass
        try:
            options.pop('MultiPV')
        except:
            pass
        try:
            options.pop('UCI_Variant')
        except:
            pass
        try:
            options.pop('UCI_Chess960')
        except:
            pass`
It was crashing because I was setting Ponder. The 2nd error was https://github.com/MarkZH/lichess-bot/blob/python-chess-upgrade/lichess-bot.py#L363 because you should take out the parentheses after move. Insetad of `move = reader.weighted_choice(board).move()` it should be `move = reader.weighted_choice(board).move`. The same should happen with the other 2 lines as well.

**Comment by MarkZH** [comment]
@AttackingOrDefending Thanks for testing. Since my engine takes very few configurations, I'll be relying on others to test my changes. I've made two more commits that should (I think) fix the bugs you found.

I'll try out version 1.4 later once I'm more confident my code is correct.

**Comment by AttackingOrDefending** [comment]
Because Ponder is managed by python-chess, we can delete it from config.yml (https://github.com/MarkZH/lichess-bot/blob/python-chess-upgrade/config.yml.default#L27). I tested it and it worked without it.

**Comment by ShailChoksi** [comment]
I will wait on more testing/confirmation before merging this in. Thanks for updating the library!

**Comment by MarkZH** [comment]
CC for testing: @AttackingOrDefending @noobpwnftw @hyperchessbot @branransom

**Comment by Cubox** [comment]
I have tested this, and it seems movetime is not being respected. Using https://github.com/ShailChoksi/lichess-bot/pull/258 made it work, but this here is not.
I'm on discord under the same name if you want me to share my configuration or run more tests

**Comment by MarkZH** [comment]
@Cubox If your bot works with #258, then that means your go_commands section was unindented one level. This branch expects the go_commands section to be indented under the uci_options as it is in the master branch. Indent the go_commands so that they are grouped under uci_options. See [lines 29-32 in this branch's config.yml.default](https://github.com/MarkZH/lichess-bot/blob/b2e18ba90975fef144bebd9617dd84908130ca10/config.yml.default#L29).

Incidentally, there was another problem. In config.yml.default, movetime is said to be specified in milliseconds, but the paramter to the chess.engine.Limit class takes seconds. So, I have to divide by 1000 [here](https://github.com/MarkZH/lichess-bot/blob/b2e18ba90975fef144bebd9617dd84908130ca10/engine_wrapper.py#L102).

**Comment by Cubox** [comment]
The issue I was having was the time not being divided by 1000.

File: engine_wrapper.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def search_with_ponder(self, board, wtime, btime, winc, binc, ponder=False):
                                         black_inc=binc / 1000,
                                         depth=cmds.get("depth"),
                                         nodes=cmds.get("nodes"),
-                                        time=cmds.get("movetime"))
+                                        time=float(cmds.get("movetime")) / 1000)
```
Comment (Cubox) [inline engine_wrapper.py:102]: if movetime is null, this will crash the bot

**Comment by MarkZH** [comment]
@Cubox Good catch. Last commit should fix that.

**Comment by Cubox** [comment]
Getting a bunch of 2021-02-14 12:48:23,210: Backing off create_engine(...) for 0.2s (AttributeError: 'Option' object has no attribute 'is_managed')

**Comment by MarkZH** [comment]
CC @zenpawn @drbeco @ddugovic for checking the Xboard protocol.

**Comment by gbtami** [comment]
I tried it with https://lichess.org/@/PyChessBot which is an Xboard variant engine. I find one issue:
```
tamas@tami:~/lichess-bot$ python3 lichess-bot.py -v
2021-02-15 16:02:08,006: 
    .   _/|
    .  // o\
    .  || ._)  lichess-bot 1.1.5
    .  //__\
    .  )___(   Play on Lichess with a bot
    
2021-02-15 16:02:08,023: Starting new HTTPS connection (1): lichess.org:443
2021-02-15 16:02:08,190: https://lichess.org:443 "GET /api/account HTTP/1.1" 200 None
2021-02-15 16:02:08,193: Welcome PyChessBot!
2021-02-15 16:02:08,194: You're now connected to https://lichess.org/ and awaiting challenges.
2021-02-15 16:02:08,245: Starting new HTTPS connection (1): lichess.org:443
2021-02-15 16:02:08,411: https://lichess.org:443 "GET /api/stream/event HTTP/1.1" 200 None
2021-02-15 16:02:23,569:     Accept Bullet casual challenge from gbtami(2056)
2021-02-15 16:02:23,575: Resetting dropped connection: lichess.org
2021-02-15 16:02:23,788: https://lichess.org:443 "POST /api/challenge/VX2om0ez/accept HTTP/1.1" 200 11
2021-02-15 16:02:23,789: --- Process Queue. Total Queued: 1. Total Used: 0
2021-02-15 16:02:23,790: --- Process Used. Total Queued: 0. Total Used: 1
2021-02-15 16:02:23,799: Starting new HTTPS connection (1): lichess.org:443
2021-02-15 16:02:23,971: https://lichess.org:443 "GET /api/bot/game/stream/VX2om0ez HTTP/1.1" 200 None
2021-02-15 16:02:23,974: Using selector: EpollSelector
2021-02-15 16:02:23,975: Using PollingChildWatcher
2021-02-15 16:02:23,981: <XBoardProtocol (pid=28938)>: Connection made
2021-02-15 16:02:23,982: <XBoardProtocol (pid=28938)>: << xboard
2021-02-15 16:02:23,982: <XBoardProtocol (pid=28938)>: << protover 2
2021-02-15 16:02:24,692: <XBoardProtocol (pid=28938)>: >> # PyChess 1.0.3 [python3 3.8.7]
2021-02-15 16:02:24,692: <XBoardProtocol (pid=28938)>: >> feature ping=1 setboard=1 playother=1 san=1 usermove=1 time=1 draw=1 sigint=0 sigterm=0 reuse=1 analyze=1 myname="PyChess 1.0.3" variants="normal,wildcastle,nocastle,fischerandom,crazyhouse,light-brigade,losers,suicide,giveaway,horde,atomic,racingkings,seirawan,kingofthehill,3check,placement,asean,cambodian,makruk,sittuyin" colors=0 ics=0 name=0 pause=0 nps=0 debug=1 memory=0 smp=0 egt="gaviota" option="skipPruneChance -slider 0 0 100"
2021-02-15 16:02:24,693: <XBoardProtocol (pid=28938)>: >> feature done=1
2021-02-15 16:02:24,693: <XBoardProtocol (pid=28938)>: Rejecting feature sigterm=0
2021-02-15 16:02:24,694: <XBoardProtocol (pid=28938)>: << rejected sigterm
2021-02-15 16:02:24,694: <XBoardProtocol (pid=28938)>: Rejecting feature usermove=1
2021-02-15 16:02:24,694: <XBoardProtocol (pid=28938)>: << rejected usermove
2021-02-15 16:02:24,694: <XBoardProtocol (pid=28938)>: Rejecting feature san=1
2021-02-15 16:02:24,694: <XBoardProtocol (pid=28938)>: << rejected san
2021-02-15 16:02:24,695: <XBoardProtocol (pid=28938)>: << accepted egt
2021-02-15 16:02:24,698: Backing off create_engine(...) for 0.2s (chess.engine.EngineError: unsupported xboard option or command: egtpath)
```
Seems the root of the problem is here https://github.com/niklasf/python-chess/blob/master/chess/engine.py#L2272
`egtpath` is not an xboard option (`egt` is it!), but a command (similar to `cores` and `smp`).

I think remove_managed_options() needs more love for Xboard engines.

**Comment by ddugovic** [comment]
Sorry I am overwhelmed but also have migrated to using the UCI-only library https://github.com/dolegi/lichess-bot so as not to depend upon Python.

**Comment by MarkZH** [comment]
@gbtami Thanks for testing. I think I see the problem. In the config.yml file, try changing

```yml
xboard_options:
  egtpath:
    gaviota: "../Gaviota/Gaviota"
```
to
```yml
xboard_options:
  egtpath gaviota: "../Gaviota/Gaviota"
```
Let me know if that works.

**Comment by gbtami** [comment]
@MarkZH that worked, thx!

**Comment by branransom** [comment]
@MarkZH I confirmed my commit is not impacted. This looks good from my end.

**Comment by MarkZH** [comment]
@gbtami Please test the latest version of the PR with the original config file. By original, I mean with this layout:

```yml
  xboard_options:            # arbitrary xboard options passed to the engine
#   cores: "4"
#   memory: "4096"
    egtpath:                 # dir containing egtb, relative to this project
      gaviota: "../Gaviota/Gaviota"
#     nalimov: "Nalimov Path"
#     scorpio: "Scorpio Path"
#     syzygy: "Syzygy Path"
```

**Comment by gbtami** [comment]
@MarkZH the latest version of PR is OK with original config now.

I faced another issue:

`2021-02-16 13:16:10,552: Backing off play_game(...) for 0.6s (FileNotFoundError: [Errno 2] No such file or directory: 'e')
`
Relevant part of my config:
```
engine:                      # engine settings
  dir: "./engines/"          # dir containing engines, relative to this project
  name: "pychess-engine"     # binary name of the engine to use
  protocol: "xboard"         # "uci" or "xboard"
  polyglot:
    enabled: true            # activate polyglot book
    book:
      standard: "engines/Perfect2017.bin" # book file path
      crazyhouse: "engines/Perfect2017.bin" # book file path
      giveaway: "engines/antichess.bin"
      atomic: "engines/atomic-2200.bin"
      horde: "engines/horde.bin"
      kingofthehill: "engines/kingOfTheHill.bin"
      racingkings: "engines/racingKings.bin"
      3check: "engines/threeCheck.bin"
    max_depth: 16             # half move max depth
    selection: "weighted_random"
```

**Comment by MarkZH** [comment]
@gbtami This problem does not involve this PR, but #236. It is expected that the books are in a list. It should work again if you change the `config.yml` from
```yml
      standard: "engines/Perfect2017.bin" # book file path
```
to
```yml
      standard:
        - "engines/Perfect2017.bin" # book file path
```

I think the code can be changed in `lichess-bot.py` to make the new multi-book lichess-bot backwards compatible with older config files. Namely, in line 369 of `lichess-bot.py`:
```python
    if isinstance(books, str):
        books = [books]

    for book in books:
        # etc.
```

**Comment by gbtami** [comment]
@MarkZH that worked, thx!

**Comment by MarkZH** [comment]
With this last bunch of commits, I'm done with this PR unless more bugs are discovered. The new python-chess library puts a lot of UCI and Xboard communication in the same interface that a lot of functions in engine_wrapper.py can be combined. However, I think I'm approaching code golf territory, so I'll stop.

Thanks again everyone who's been testing this out!

**Pull request merged**